src/test/unittest/context.py

```diff
@@ -76,24 +78,59 @@ def create_holey_file(self, size, name):
 
     def exec(self, cmd, args='', expected_exit=0):
         """Execute binary in current test context"""
+        if expected_exit == 0:
+            self._normal_exit_init()
+        else:
+            self._abnormal_exit_init()
+
         env = {**self.env, **os.environ.copy(), **self.test.utenv}
         if sys.platform == 'win32':
             env['PATH'] = self.build.libdir + os.pathsep + env.get('PATH', '')
             cmd = os.path.join(self.build.exedir, cmd) + '.exe'
         else:
+            env['LD_PRELOAD'] = env.get('LD_PRELOAD', '') + os.pathsep +\
+                self.test.ld_preload
             env['LD_LIBRARY_PATH'] = self.build.libdir + os.pathsep +\
-                                     env.get('LD_LIBRARY_PATH', '')
+                env.get('LD_LIBRARY_PATH', '')
             cmd = os.path.join(self.test.cwd, cmd) + self.build.exesuffix
+            cmd = '{} {}'.format(self.valgrind.cmd, cmd)
 
-        proc = sp.run([cmd, args], env=env, cwd=self.test.cwd,
-                      timeout=self.conf.timeout, stdout=sp.PIPE,
+        proc = sp.run('{} {}'.format(cmd, args), env=env, cwd=self.test.cwd,
+                      shell=True, timeout=self.conf.timeout, stdout=sp.PIPE,
                       stderr=sp.STDOUT, universal_newlines=True)
 
+        if expected_exit == 0 and not self.valgrind.validate_log():
+            self.test.fail(proc.stdout)
+
         if proc.returncode != expected_exit:
             self.test.fail(proc.stdout)
         else:
             self.msg.print_verbose(proc.stdout)
 
+    def _normal_exit_init(self):
+        """
+        Set configurations suitable for expecting normal exit from process
+        """
+        if self.valgrind.tool == vg._Tool.HELGRIND:
+            self.valgrind.add_suppression('helgrind-log.supp')
+        elif self.valgrind.tool == vg._Tool.MEMCHECK:
+            self.valgrind.add_suppression('memcheck-dlopen.supp')
+        elif self.valgrind.tool == vg._Tool.DRD:
+            self.valgrind.add_suppression('drd-log.supp')
+
+        if self.test.ld_preload:
+            self.valgrind.handle_ld_preload(self.test.ld_preload)
+
+    def _abnormal_exit_init(self):
+        """
+        Set configurations suitable for expecting abnormal exit from process
+        """
+        if self.valgrind.tool == vg._Tool.DRD:
+            self.valgrind.add_suppresssion('drd-log.supp')
```

Comment:
I think this is actually a bug in shell's expect_abnormal_exit.
Please share the setup code between expect_normal_exit and expect_abnormal_exit.